### PR TITLE
Dex Drag-In Update:

### PR DESCRIPTION
**Pull request by GSV-a-Nest-of-Kindled-Embers**
Dex Drag-In Update:
- Dex Drag-In now places generated actors in folders named after the scene they were generated in, to avoid filling up one single folder over time and ease organization.
- Added Level Range (min and max), Shiny Chance, Stat Randomness, and Prevent Evolution options to the dex drag-in popup, as well as config options to set their defaults (except prevent evolution).
- Added the ability for dex drag-in to generate and apply shiny status/sprites.

### CODE REVIEW

File: module/forms/dex-drag-options-form.js [hunk 1]
```diff
@@ -25,15 +26,33 @@ export class PTUDexDragOptions extends FormApplication {
...
       return data;
     }
 
     /* -------------------------------------------- */
     
     /** @override */
     async _updateObject(event, formData) {
-        console.log("formData");
-        console.log(formData);
+        debug("formData");
```
Comment (dylanpiera) [inline module/forms/dex-drag-options-form.js:48]:
```suggestion
```

File: module/forms/dex-drag-options-form.js [hunk 2]
```diff
@@ -25,15 +26,33 @@ export class PTUDexDragOptions extends FormApplication {
...
+        debug(formData);
```
Comment (dylanpiera) [inline module/forms/dex-drag-options-form.js:49]:
```suggestion
```

File: module/utils/actor-generator.js [hunk 3]
```diff
@@ -92,6 +99,30 @@ export class ActorGenerator {
...
+        let is_shiny = false;
+        if(options.shiny_chance)
+        {
+            let shiny_roll = getRandomIntInclusive(1, 100);
+
+            if(shiny_roll <= options.shiny_chance)
+            {
+                is_shiny = true;
+            }
+        }
```
Comment (dylanpiera) [inline module/utils/actor-generator.js:112]:
```suggestion

        const is_shiny = options?.shiny_chance ? (getRandomIntInclusive(1, 100) <= options.shiny_chance) : false;
```

File: module/utils/actor-generator.js [hunk 4]
```diff
@@ -92,6 +99,30 @@ export class ActorGenerator {
...
+
+        let stat_randomness = 20;
+        if(options.stat_randomness)
+        {
+            stat_randomness = options.stat_randomness;
+        }
```
Comment (dylanpiera) [inline module/utils/actor-generator.js:118]:
```suggestion
        const stat_randomness = options?.stat_randomness ?? 20;
```

File: module/utils/actor-generator.js [hunk 5]
```diff
@@ -92,6 +99,30 @@ export class ActorGenerator {
...
+
+        let prevent_evolution = false;
+        if(options.prevent_evolution)
+        {
+            prevent_evolution = true;
+        }
```
Comment (dylanpiera) [inline module/utils/actor-generator.js:124]:
```suggestion
        const prevent_evolution = options?.prevent_evolution ?? false;
```

File: module/utils/actor-generator.js [hunk 6]
```diff
@@ -113,8 +144,8 @@ export class ActorGenerator {
         await GetOrCacheAbilities(); await GetOrCacheMoves(); await GetOrCacheCapabilities();
 
         if(options.exists) return ag;
-        
-        await ag.Generate().PrepareArt(options.imgpath ?? undefined);
+
+        await ag.Generate("weighted", undefined, undefined, undefined, stat_randomness, is_shiny, prevent_evolution).PrepareArt(options.imgpath ?? undefined, undefined, is_shiny);
```
Comment (dylanpiera) [inline module/utils/actor-generator.js:148]: Okay this is definitely gonna need refactoring. No way I'm allowing triple undefineds to plague the codebase 😂 

Will need to change `Generate` to take in a destructured list of parameters instead, will set this up when I actually do testing of this codebase.

File: module/utils/actor-generator.js [hunk 7]
```diff
@@ -132,7 +163,12 @@ async function ApplyChanges() {
     return this.actor;
 }
 
-function PrepareEvolution() {
+function PrepareEvolution(prevent_evolution = undefined) {
+    if(prevent_evolution)
+    {
+        return this;
+    }
```
Comment (dylanpiera) [inline module/utils/actor-generator.js:170]:
```suggestion
    if(prevent_evolution) return this;
```

File: module/utils/actor-generator.js [hunk 8]
```diff
@@ -267,4 +303,9 @@ function PrepareCapabilities(allCapabilities) {
     
     this.items = this.items.concat(newCapabilities.filter(x => x !== undefined));
     return this;
+}
+
+function PrepareShiny(is_shiny = false) {
```
Comment (dylanpiera) [inline module/utils/actor-generator.js:308]: Should this function not just take a percentage and then calculate whether a pokemon should be shiny. So that piece of code can be moved from the create method over here instead. @GSV-a-Nest-of-Kindled-Embers

File: module/utils/species-command-parser.js [hunk 9]
```diff
@@ -79,21 +79,27 @@ export async function CreateMonParser(input, andCreate = false) {
...
-export async function GetSpeciesArt(mon, imgDirectoryPath, type = ".png") {
+export async function GetSpeciesArt(mon, imgDirectoryPath, type = ".png", shiny = false) {
     const basePath = imgDirectoryPath+(imgDirectoryPath.endsWith('/') ? '' : '/')
 
-    let path = basePath+lpad(mon?.number, 4)+type;
+    let shiny_path = "";
+    if(shiny)
+    {
+        shiny_path = "s";
+    }
```
Comment (dylanpiera) [inline module/utils/species-command-parser.js:89]:
```suggestion
    const shiny_path = shiny ? "s" : "";
```
Not using null coalescing operator here because realitstically the 'shiny' path could be `""` or the default `false`.

Should maybe instead have this shiny path be set from a variable, like 
```js
const shiny_path = shiny === true ? "s" : shiny ? shiny : "";
```

File: module/utils/species-command-parser.js [hunk 10]
```diff
@@ -164,7 +170,7 @@ Hooks.on("dropCanvasData", (canvas, update) => {
         if(item) update.item = item;
     }
     if(update.item)
-        new game.ptu.PTUDexDragOptions(update, {"submitOnChange": false, "submitOnClose": true}).render(true);
+        new game.ptu.PTUDexDragOptions(update, {"submitOnChange": false, "submitOnClose": false}).render(true);
```
Comment (dylanpiera) [inline module/utils/species-command-parser.js:173]: @GSV-a-Nest-of-Kindled-Embers why this change? if a user presses enter/esc/close on this box it is most likely to skip the options, not to cancel generation. 

```suggestion
        new game.ptu.PTUDexDragOptions(update, {"submitOnChange": false, "submitOnClose": true}).render(true);
```

**Review by dylanpiera — CHANGES_REQUESTED** [review]
Hey Void, I just had a quick look over *purely* based on what you programmed, left feedback (both on the way you wrote the code, as well as in some places the functionality) and cleaned up some of your code.

Please take a look at the comments and make the one change I suggested in relate to the `PrepareShiny` function.

Besides that I'll try to sometime soon make the change I'm talking about in my comments, and actually give testing the code a shot. But as I said earlier currently on vacation so might take me a bit longer.

Cheers, and thanks for the work!

File: module/utils/species-command-parser.js (hunk 10 above)
Comment (GSV-a-Nest-of-Kindled-Embers) [reply module/utils/species-command-parser.js:169]: When I was working on this and testing, I closed the popup and saw that it still generated, and at first I worried that _I_ had introduced a bug somehow, and afterwards just assumed it was just a pre-existing bug. Hitting close or ESC and the operation not cancelling is definitely not what I would think of as expected behavior, IMO.

File: module/utils/actor-generator.js (hunk 6 above)
Comment (GSV-a-Nest-of-Kindled-Embers) [reply module/utils/actor-generator.js:131]: Yeah, sorry about that - I was sort of snaking values through there in an effort to avoid changing anything else that might call that function, but a refactor would definitely be cleaner at this point.

File: module/utils/actor-generator.js (hunk 8 above)
Comment (GSV-a-Nest-of-Kindled-Embers) [reply module/utils/actor-generator.js:288]: That sounds sensible, yeah.

File: module/utils/species-command-parser.js (hunk 10 above)
Comment (GSV-a-Nest-of-Kindled-Embers) [reply module/utils/species-command-parser.js:169]: I _do_ want users to be able to just hit **Enter** and skip the options to continue generation the already-filled defaults, but I'm not sure I was able to get that working consistently with the `autofocus` tag on the submit button.

File: module/utils/actor-generator.js (hunk 8 above)
Comment (GSV-a-Nest-of-Kindled-Embers) [reply module/utils/actor-generator.js]: Cleaned up in this commit: https://github.com/GSV-a-Nest-of-Kindled-Embers/Foundry-Pokemon-Tabletop-United-System/commit/23bb9126a0f91f0cc9a3ec52bcdbbe326eb60b66

File: module/utils/actor-generator.js [hunk 11]
```diff
@@ -70,10 +71,16 @@ export class ActorGenerator {
...
             else throw 'Capabilities not cached, please provide moves.';
         }
-        return this.PrepareEvolution().PrepareGender().PrepareStats(statMethod).PrepareMoves(allMoves).PrepareAbilities(allAbilities).PrepareCapabilities(allCapabilities);
+
+        let stat_rng = 0.1;
+        if(stat_randomness)
+        {
+            stat_rng = Number(stat_randomness/100);
+        }
+        return this.PrepareEvolution(prevent_evolution).PrepareGender().PrepareStats(statMethod, stat_rng).PrepareMoves(allMoves).PrepareAbilities(allAbilities).PrepareCapabilities(allCapabilities).PrepareShiny(shiny_chance);
```
Comment (dylanpiera) [inline module/utils/actor-generator.js:80]: Unnecessary variables again, could just make use of stat_randomness, also we have a default value in the settings so why do we have hardcoded 0.1

Will fix this in local as it won't let me select all lines at ones on the github website.

File: module/utils/actor-generator.js [hunk 12]
```diff
@@ -113,8 +125,8 @@ export class ActorGenerator {
         await GetOrCacheAbilities(); await GetOrCacheMoves(); await GetOrCacheCapabilities();
 
         if(options.exists) return ag;
-        
-        await ag.Generate().PrepareArt(options.imgpath ?? undefined);
+
+        await ag.Generate("weighted", undefined, undefined, undefined, stat_randomness, options.shiny_chance, prevent_evolution).PrepareArt(options.imgpath ?? undefined);
```
Comment (dylanpiera) [inline module/utils/actor-generator.js:129]: note to self: I really need to rework this. lmao

**Review by dylanpiera — APPROVED** [review]

**Comment by dylanpiera** [comment]
Making some small fixes in local and will then merge this.
